File: code/ch09/javascript/median_property.js
// §9.1.4 Test Oracles — property oracle for median(xs), as a property-based
// test. Uses fast-check when installed (npm install fast-check); otherwise
// falls back to a hand-rolled random-input loop over the same two properties.

function median(xs) {
  const ys = xs.toSorted((a, b) => a - b);
  return ys[Math.floor(xs.length / 2)];   // middle element (upper median)
}

function propertiesHold(xs) {
  const m = median(xs);
  if (m !== median(xs.toReversed())) throw new Error(`order dependence on ${xs}`);
  if (!xs.includes(m)) throw new Error(`median ${m} not in ${xs}`);
}

let fc = null;
try { fc = require("fast-check"); } catch { /* not installed: use the fallback */ }

if (fc) {
  fc.assert(
    fc.property(fc.array(fc.integer(), { minLength: 1 }), (xs) => {
      propertiesHold(xs);
      return true;
    })
  );
  console.log("fast-check: both properties held on generated lists");
} else {
  for (let i = 0; i < 10000; i++) {
    const n = 1 + Math.floor(Math.random() * 50);
    const xs = Array.from({ length: n }, () => Math.floor(Math.random() * 2001) - 1000);
    propertiesHold(xs);
  }
  console.log("10000 random lists: both properties held");
}

File: code/ch09/ruby/median_property.rb
# §9.1.4 Test Oracles — property oracle for median(xs), as a property-based
# test. Uses PropCheck when installed (gem install prop_check); otherwise
# falls back to a hand-rolled random-input loop over the same two properties.

def median(xs)
  xs.sort[xs.length / 2]                     # middle element (upper median)
end

def properties_hold(xs)
  m = median(xs)
  raise "order dependence on #{xs}" unless m == median(xs.reverse)
  raise "median #{m} not in inputs" unless xs.include?(m)
end

begin
  require "prop_check"
  g = PropCheck::Generators
  PropCheck.forall(g.array(g.integer, empty: false)) { |xs| properties_hold(xs) }
  puts "PropCheck: both properties held on generated lists"
rescue LoadError
  srand(9)
  10_000.times do
    xs = Array.new(rand(1..50)) { rand(-1000..1000) }
    properties_hold(xs)
  end
  puts "10000 random lists: both properties held"
end
